File: app/src/main/java/com/kira/kdownloader/util/RecentUrls.java
package com.kira.kdownloader.util;

import android.content.Context;
import android.content.SharedPreferences;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public final class RecentUrls {
    private static final String PREFS = "recent_urls";
    private static final String KEY = "urls";
    private static final int MAX = 3;

    private RecentUrls() {}

    public static List<String> all(Context context) {
        String stored = prefs(context).getString(KEY, "");
        if (stored == null || stored.isEmpty()) return Collections.emptyList();
        List<String> result = new ArrayList<>();
        for (String value : stored.split("\\n")) {
            if (!value.trim().isEmpty()) result.add(value);
            if (result.size() == MAX) break;
        }
        return result;
    }

    public static void add(Context context, String url) {
        String trimmed = url.trim();
        if (trimmed.isEmpty()) return;
        List<String> updated = new ArrayList<>();
        updated.add(trimmed);
        for (String value : all(context)) {
            if (!trimmed.equals(value) && updated.size() < MAX) updated.add(value);
        }
        prefs(context).edit().putString(KEY, join(updated)).apply();
    }

    public static void remove(Context context, String url) {
        List<String> updated = new ArrayList<>();
        for (String value : all(context)) if (!url.equals(value)) updated.add(value);
        prefs(context).edit().putString(KEY, join(updated)).apply();
    }

    private static String join(List<String> values) {
        StringBuilder result = new StringBuilder();
        for (String value : values) {
            if (result.length() > 0) result.append('\n');
            result.append(value);
        }
        return result.toString();
    }

    private static SharedPreferences prefs(Context context) {
        return context.getApplicationContext().getSharedPreferences(PREFS, Context.MODE_PRIVATE);
    }
}

File: app/src/main/java/com/kira/kdownloader/util/RecentUrls.kt
package com.kira.kdownloader.util

import android.content.Context

/**
 * Small persistent list of recently fetched URLs, surfaced as quick-pick chips on the Home idle
 * screen. Backed by SharedPreferences; most-recent first, de-duplicated, capped at [MAX].
 */
object RecentUrls {
    private const val PREFS = "recent_urls"
    private const val KEY = "urls"
    private const val SEP = "\n"
    private const val MAX = 8

    fun all(context: Context): List<String> =
        prefs(context).getString(KEY, "").orEmpty()
            .split(SEP)
            .filter(String::isNotBlank)

    fun add(context: Context, url: String) {
        val trimmed = url.trim()
        if (trimmed.isBlank()) return
        val updated = (listOf(trimmed) + all(context).filter { it != trimmed }).take(MAX)
        prefs(context).edit().putString(KEY, updated.joinToString(SEP)).apply()
    }

    fun remove(context: Context, url: String) {
        val updated = all(context).filter { it != url }
        prefs(context).edit().putString(KEY, updated.joinToString(SEP)).apply()
    }

    private fun prefs(context: Context) =
        context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
}
